### PR TITLE
Merge multiple clients into single client

### PR DESCRIPTION
**Pull request by kasthuriraajan**
## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

**Comment by LakshanSS** [comment]
@kasthuriraajan Shall we add the display annotations also in this PR?
I added them in https://github.com/ballerina-platform/module-ballerinax-mongodb/pull/90
But that PR is outdated. So I will close it.
Shall we add it with this PR? What do you think?

**Comment by abeykoon** [comment]
This PR is for merging the clients so that Choreo UI can render the configurations. Let us keep annotations PR separate.

**Comment by LakshanSS** [comment]
> This PR is for merging the clients so that Choreo UI can render the configurations. Let us keep annotations PR separate.

+1

**Comment by abeykoon** [comment]
@kasthuriraajan shall we check why Travis build has failed? Seems there is a dependency issue.

File: mongodb/client.bal [hunk 1]
```diff
@@ -21,51 +21,188 @@ import ballerina/jballerina.java;
...
+
+    # Deletes a document based on a condition.
+    # 
+    # + databaseName - Name of the database 
+    # + collectionName - Name of the collection
+    # + filter - Filter for the query
+    # + isMultiple - Delete multiple documents if the condition is matched
+    # + return - The number of deleted documents or else a `mongodb:Error` if unable to reach the DB
+    remote function delete(string collectionName, string? databaseName = (),map<json>? filter = (), boolean isMultiple = false)
+     returns int|Error {
```
Comment (RolandHewage) [inline mongodb/client.bal:180]: Check code formatting

File: mongodb/client.bal [hunk 2]
```diff
@@ -21,51 +21,188 @@ import ballerina/jballerina.java;
...
     # Closes the client.
-    # ```ballerina
-    # mongoClient->close();
-    # ```
     remote function close() {
         close(self.datasource);
     }
+
+    function getCurrentDatabase(string? databaseName) returns handle|Error {
+        if(databaseName is string) {
```
Comment (RolandHewage) [inline mongodb/client.bal:195]: Add a single space after 'if'

File: mongodb/client.bal [hunk 3]
```diff
@@ -21,51 +21,188 @@ import ballerina/jballerina.java;
...
+            handle database = check self.getDatabase(databaseName);
+            return database;
+        } else {
+            if(!java:isNull(self.database)) {
```
Comment (RolandHewage) [inline mongodb/client.bal:199]: Add a single space after 'if'

File: mongodb/client.bal [hunk 4]
```diff
@@ -21,51 +21,188 @@ import ballerina/jballerina.java;
...
+
+    // Collection service operations
+    # Counts the documents based on the filter. When the filter is (), it counts all the documents in the collection.
+    #
+    # + databaseName - Name of the database 
+    # + collectionName - Name of the collection
+    # + filter - Filter for the count ($where & $near can be used)
+    # + return - Count of the documents in the collection or else `mongodb:Error` if unable to reach the DB
+    remote function countDocuments(string collectionName, string? databaseName = (), map<json>? filter = ()) 
+    returns int|Error {
```
Comment (RolandHewage) [inline mongodb/client.bal:94]: Bring the return statement inline with the staring bracket of the method signature

File: mongodb/client.bal [hunk 5]
```diff
@@ -21,51 +21,188 @@ import ballerina/jballerina.java;
...
+    # The queries collection for documents, which sorts and limits the returned results.
+    #
+    # + databaseName - Name of the database 
+    # + collectionName - Name of the collection
+    # + filter - Filter for the query
+    # + sort - Sort options for the query
+    # + limit - Limit options for the query results. No limit is applied for -1
+    # + return - JSON array of the documents in the collection or else a `mongodb:Error` if unable to reach the DB
+    remote function find(string collectionName, string? databaseName = (),map<json>? filter = (), map<json>? sort = (), 
+    int 'limit = -1) returns map<json>[]|Error {
```
Comment (RolandHewage) [inline mongodb/client.bal:135]: Bring the parameter inline with the starting bracket of the method signature

File: mongodb/client.bal [hunk 6]
```diff
@@ -21,51 +21,188 @@ import ballerina/jballerina.java;
...
+    #
+    # + databaseName - Name of the database 
+    # + collectionName - Name of the collection
+    # + set - Document for the update condition
+    # + filter - Filter for the query
+    # + isMultiple - Whether to update multiple documents
+    # + upsert - Whether to insert if update cannot be achieved
+    # + return - JSON array of the documents in the collection or else a `mongodb:Error` if unable to reach the DB
+    remote function update(map<json> set,string collectionName, string? databaseName = (), map<json>? filter = (), 
+    boolean isMultiple = false, boolean upsert = false) returns int|Error {
```
Comment (RolandHewage) [inline mongodb/client.bal:162]: Bring the parameter inline with the starting bracket of the method signature

File: mongodb/client.bal [hunk 7]
```diff
@@ -84,6 +221,43 @@ function close(handle datasource) = @java:Method {
...
+function listIndices(handle collection) returns map<json>[]|DatabaseError = @java:Method {
+    'class: "org.wso2.mongo.MongoDBCollectionUtil"
+} external;
+
+function insert(handle collection, handle document) returns DatabaseError? = @java:Method {
+    'class: "org.wso2.mongo.MongoDBCollectionUtil"
+} external;
+
+function find(handle collection, handle? filter, handle? sort, int 'limit)
+returns map<json>[]|DatabaseError = @java:Method {
```
Comment (RolandHewage) [inline mongodb/client.bal:248]: Bring the parameter inline with the starting bracket of the method signature

File: mongodb/client.bal [hunk 8]
```diff
@@ -84,6 +221,43 @@ function close(handle datasource) = @java:Method {
...
+    'class: "org.wso2.mongo.MongoDBCollectionUtil"
+} external;
+
+function update(handle collection, handle update, handle? filter, boolean isMultiple, boolean upsert)
+returns int|DatabaseError = @java:Method {
```
Comment (RolandHewage) [inline mongodb/client.bal:253]: Bring the parameter inline with the starting bracket of the method signature

**Review by RolandHewage — COMMENTED** [review]
Few formatting issues to be solved

**Review by RolandHewage — APPROVED** [review]

**Review by abeykoon — APPROVED** [review]

File: README.md [hunk 9]
```diff
@@ -1,41 +1,420 @@
...
+```ballerina
+mongodb:ClientConfig mongoConfig = {
+        host: host,
+        port: port,
+        username: username,
+        password: password,
+        options: {sslEnabled: false, serverSelectionTimeout: 5000}
+    };
+
+    mongodb:Client mongoClient = checkpanic new (mongoConfig, database);
```
Comment (indikasampath2000) [inline README.md:84]: Please provide commit or PR if `checkpanic` already has been fixed in the master?

Comment (kasthuriraajan) [reply README.md:84]: This code snippet is now removed from here and added in Module.md file. This change is done via the following PR https://github.com/ballerina-platform/module-ballerinax-mongodb/pull/111 . You can see the change in Module.md https://github.com/ballerina-platform/module-ballerinax-mongodb/pull/111/files#diff-db8c92970e0763b610645268a9105ee93f96ad2fcae5c68d92d348634b27aa77R38 here.

**Pull request merged**